### PR TITLE
Add guide for programmatically creating pages

### LINKED ISSUE #8841: Content migration - fill out "Programmatically create pages from data" stub article in the docs
## Issue Context

This issue is part of the content migration project (#8103), which is meant to take existing content and *repurpose* it so it's easier to find in the docs.

### Motivation

We are doing this because many people have written excellent content that we want to canonize and maintain so more people can find it and benefit from it.

### Review the Gatsby docs approach

- [ ] Read the Style Guide and follow it in order for your PR to be accepted. Pay special attention to the templates. 
  - Style Guide: https://www.gatsbyjs.org/docs/gatsby-style-guide/
  - Docs Templates: https://www.gatsbyjs.org/docs/templates/
  - Example of a similar article: [Adding Markdown Pages
](https://www.gatsbyjs.org/docs/adding-markdown-pages/)

### Draft the doc

- [ ] Add to the [Programmatically create pages from data](https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/programmatically-create-pages-from-data.md) stub article.
- [ ] Use the following reference material as a primary starting point to draft the new doc: https://www.gatsbyjs.org/tutorial/part-seven/ Make sure you rewrite and adjust the content to follow the format listed in these resources:
  - Docs Templates: https://www.gatsbyjs.org/docs/templates/
  - Example of a guide to imitate: https://www.gatsbyjs.org/docs/deploying-to-s3-cloudfront
- [ ] Find the stub in the [docs sidebar](https://github.com/gatsbyjs/gatsby/blob/master/www/src/data/sidebars/doc-links.yaml), and remove the asterisk next to the title, (which will indicate that the doc is no longer a stub).

### Open a pull request
- [ ] Open a pull request with your work, including the words "closes #8841"

### PR DESCRIPTION
**Pull request by jbranchaud**
closes #8841

### CODE REVIEW

**Comment by jbranchaud** [comment]
This is still a work-in-progress, but I wanted to show that it is being worked on.

**Comment by jbranchaud** [comment]
It is now ready for review/feedback.

**Comment by pieh** [comment]
@jlengstorf do we want to mention (and show) how to create pages using "unstructured data" here?  Or would it be completely separate?

File: docs/docs/programmatically-create-pages-from-data.md [hunk 1]
```diff
@@ -2,7 +2,108 @@
...
+        });
+      });
+      resolve();
+    });
+  });
+};
+```
+
+For each page we want to create we must specify the `path` for visiting that
+page, the `component` template used to render that page, and any `context`
```
Comment (pieh) [inline docs/docs/programmatically-create-pages-from-data.md]: maybe mention that `context` is optional? By default gatsby will pass some context already - stuff like `path`

File: docs/docs/programmatically-create-pages-from-data.md [hunk 2]
```diff
@@ -2,7 +2,108 @@
...
+      html
+      frontmatter {
+        title
+      }
+    }
+  }
+`;
+```
+
+Notice that the `slug` value we specified in the `createPage` context is
```
Comment (pieh) [inline docs/docs/programmatically-create-pages-from-data.md]: We can also use context directly in template component - it's accessible via `pageContext` prop

**Comment by amberleyromo** [comment]
@pieh @jlengstorf I think we could link to the unstructured data doc at the end so people could hop over if they're interested. https://www.gatsbyjs.org/docs/using-unstructured-data/

**Comment by amberleyromo** [comment]
Hi @jbranchaud! Following up -- want to make these edits?

**Comment by jbranchaud** [comment]
@amberleyromo Thanks for the followup! I've made the edits. Let me know how it looks.

**Comment by jbranchaud** [comment]
@pieh @amberleyromo how does this PR look? Are there any other edits needed?

**Pull request merged**